src/Datadog.Trace/Sampling/RuleBasedSampler.cs

```diff
@@ -0,0 +1,134 @@
+using System;
+using System.Collections.Generic;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Sampling
+{
+    internal class RuleBasedSampler : ISampler
+    {
+        private const ulong KnuthFactor = 1_111_111_111_111_111_111;
+
+        private static readonly Vendors.Serilog.ILogger Log = DatadogLogging.For<RuleBasedSampler>();
+
+        private readonly IRateLimiter _limiter;
+        private readonly List<ISamplingRule> _rules = new List<ISamplingRule>();
+
+        private readonly object _sampleRateGate = new object();
+
+        private Dictionary<string, float> _sampleRates = new Dictionary<string, float>();
+
+        public RuleBasedSampler(IRateLimiter limiter)
+        {
+            _limiter = limiter ?? new RateLimiter(null);
+        }
+
+        public void SetDefaultSampleRates(IEnumerable<KeyValuePair<string, float>> sampleRates)
+        {
+            // to avoid locking if writers and readers can access the dictionary at the same time,
+            // build the new dictionary first, then replace the old one
+            var rates = new Dictionary<string, float>(StringComparer.OrdinalIgnoreCase);
+
+            if (sampleRates != null)
```

Comment:
Seems like we shouldn't allow null `sampleRates`. Should we throw a `ArgumentNullException` instead?

Reply:
This is also carry over.